### PR TITLE
Resolve "Unable to immediately select pre-selected login"

### PR DESCRIPTION
**Pull request by cognophile**
Hey! First-time contributor to any OS project so I hope this PR can provide some value to the application and I'm giving enough detail and such. I'm enjoying exploring this project so any feedback and suggestions you might have would be appreciated!

I've not purchased the application on the app-store, so I can't confirm whether these issues are present in the shipped version, however, once my environment was configured, both impeded me from using the application. 

This pull request contains: 
- A fix for #4. I added a new `TableView` extension to enable the detection of clicks on the pre-selected top-item in the left-hand login list. 
- A fix to enable authentication with the application when no logins are stored (eg. first use [one half of #3]) by guarding against a `0` count return from the `login` table. I'm wondering if recent changes in `SQLite.swift` might have influenced this? 
- Grouped some files together for navigational clarity. Not sure about one or two of the names, so feel free to voice any preferences on these, or if no groups at all is preferred! 

Here is the fix for #4 in action! 
FixedTableViewSelectionBug

### CODE REVIEW

**Comment by swlkr** [comment]
Hey! This is great! Way more than I could have asked for haha.

Um, so I loaded up your fork in xcode and I get #3 but I don't get #4.

Is it supposed to highlight the first one and show it in the right pane?

**Comment by cognophile** [comment]
@swlkr No problem! Thanks for taking the time to check this out - really appreciate it! 

Oh, that's strange. I downloaded a fresh copy of the project at `383946f` to sanity check and ensure it wasn't just my fork, and both #3 and #4 are still present. 

My understanding was that (by design?) once the user has authenticated, the first login item should be highlighted in the deep grey colour, indicating a kind of pre-selection, but I guessed your intention was that it shouldn't be loaded into the right-hand pane until clicked? The problem is that if you then click on that first item, it doesn't fire the `tableViewSelectionDidChange` function. My theory is that because the code that applies the grey box in `reloadTableView` on-load assigns `index` (which has a default value of `0`) to the `tableView.selectedItem`. As the item at the top of the list is at index `0`, when clicked to load it in the right pane, I'm thinking the selection doesn't change because the `selectedItem` value doesn't? This seems to be confirmed if selecting any login other than the `0`th one, and then the `0`th one after. 

So I thought I'd have a got at applying the fix in this PR to allow the first item to be shown in the right-hand pane when it's the first item clicked, after view load, but still preserve the default grey highlighting to indicate it's a field the user can interact with. 

**Update**: 
I'm still looking into #3. I'm not sure the guard solution is suitable as it prevents the rest of `ContainerViewController::viewWillAppear()` from running - eg. all lines after 
```
let rowCount = try? db?.scalar(login.table.count)
```

**Comment by cognophile** [comment]
I've updated the issue (#3) with my renewed findings. I've also updated this PR to reflect this by:

- Removed the inserted `guard` statements from `EditViewController::cancelButtonClicked` and `ContainerViewController::viewWillAppear()`. Both are reset to their original working `let rowCount = try? db?.scalar(login.table.count)` statements.
- Updated the `Podfile` and `Pods/` to utilise the latest `master` commit of `SQLite.swift`. I've followed the previous convention of setting the reference to the commit id assuming there is a particular desire for this (such as preventing auto-updates?). However, moving from the commit id 
```
pod 'SQLite.swift/SQLCipher', :git => 'https://github.com/stephencelis/SQLite.swift.git', :commit => '1a908a7da11852f252e7c6b6366a4d9f8a7d5272'
```
to the branch reference is probably a safer strategy, should the git history of the branch be re-written (eg. rebases)
```
pod 'SQLite.swift/SQLCipher', :git => 'https://github.com/stephencelis/SQLite.swift.git', :branch => 'master'
```
- Further improved the file grouping structure.
- Update the table view (login list) keyboard shortcuts to be more intuitive for email, password, and delete.

File: allyourpasswords/Views/ToolbarCustomView.swift [hunk 1]
```diff
@@ -9,45 +9,45 @@
 import Cocoa
 
 class ToolbarCustomView: NSView {
-
+
```
Comment (swlkr) [inline allyourpasswords/Views/ToolbarCustomView.swift]: What's up with all of the whitespace changes in this file?

**Comment by swlkr** [comment]
Wow this PR really grew in size, I'm usually pretty laid back about this stuff, and I'll probably accept it anyway but I guess for the future, it's better to have small PRs, so with this one as an example:

- a PR to move the files around (which… I mean really is kind of aesthetic)
- another PR to change SQLCipher from the sha to the branch
- another PR to handle the empty db issues
- one more PR to handle the selection problem on startup

So don't change it now, we'll just leave it in the one big PR

**Comment by swlkr** [comment]
Alright so two things:

1. sqlite.swift from SHA to master

We're pretty much trading one problem for another. The SHA will break if there's a rebase, but with it pointing to master, things will definitely be changing on every pod install any time anyone pushes to master in that repo… I mean I don't have a solution, just wanted to point out the trade off

2. The first item selection problem

I was messing around with this and I couldn't seem to get the first item to be selected programmatically. So it's not that I wanted it to be highlighted in gray, I just settled because of the month long time constraint.

File: allyourpasswords/Views/EditViewController.swift [hunk 2]
```diff
@@ -138,7 +138,7 @@ class EditViewController : NSViewController {
     @IBAction func cancelButtonClicked(_ sender: NSButton) {
         let container = self.parent as! ContainerViewController
         let rowCount = try! db?.scalar(login.table.count)
-
+
```
Comment (swlkr) [inline allyourpasswords/Views/EditViewController.swift]: Is this a tabs vs spaces situation? Should probably settle on something here

Comment (swlkr) [inline allyourpasswords/Views/EditViewController.swift]: Oh yeah ok so this whitespace is just trailing, be sure in the future to delete the trailing whitespace, especially on blank lines

**Comment by swlkr** [comment]
Overall it looks good! I'm happy to accept if this PR is in a steady state

File: allyourpasswords/Views/ToolbarCustomView.swift (hunk 1 above)
Comment (cognophile) [reply allyourpasswords/Views/ToolbarCustomView.swift]: Hmm. Not sure - sorry! I'll fix this up.

File: allyourpasswords/Views/EditViewController.swift (hunk 2 above)
Comment (cognophile) [reply allyourpasswords/Views/EditViewController.swift]: Will do - thanks! I'll remove it.

**Comment by cognophile** [comment]
Thanks, Sean! Appreciate you having taken the time to look through this and pass on some of your expertise. I'll definitely keep your feedback in mind for the future - valuable stuff 🙂  

> Wow this PR really grew in size, I'm usually pretty laid back about this stuff, and I'll probably accept it anyway but I guess for the future, it's better to have small PRs, so with this one as an example:
> * a PR to move the files around (which… I mean really is kind of aesthetic)
> * another PR to change SQLCipher from the sha to the branch
> * another PR to handle the empty db issues
> * one more PR to handle the selection problem on startup

Hmm, yes, I see how that could be problematic. Sorry about that. Thanks for the feedback - I'm still getting the hang of OS etiquette outside of single-author projects 🙂 

> sqlite.swift from SHA to master
> We're pretty much trading one problem for another. The SHA will break if there's a rebase, but with it pointing to master, things will definitely be changing on every pod install any time anyone pushes to master in that repo… I mean I don't have a solution, just wanted to point out the trade off

I agree. It's a tough one. It could reference a git tag, but of course, the rebase issue would still be possible should the SHA behind the tag change. But, it would avoid the rolling-updates of relying on a branch, and hopefully avoid rebase issues if the pod maintainers had a method of preserving tags if they wished to rebase but there is no guarantee. Eg.
```
pod 'Foo', :git => 'https://github.com/Bar/Foo.git', :tag => '1.2.3'
```
I'll leave it as the SHA for now. 

> I was messing around with this and I couldn't seem to get the first item to be selected programmatically. So it's not that I wanted it to be highlighted in gray, I just settled because of the month long time constraint.

Ah, I see. I've got the first item loading programmatically when first loading the application after login and is highlighted grey indicating a non-user selection.

I'll just address the whitespace issues as I didn't realise they'd snuck in!

File: allyourpasswords/Views/ContainerViewController.swift [hunk 3]
```diff
@@ -26,9 +26,8 @@ class ContainerViewController : NSViewController {
 
         db = Database.open()
         let login = Login()
-
         let rowCount = try! db?.scalar(login.table.count)
-
+
```
Comment (cognophile) [inline allyourpasswords/Views/ContainerViewController.swift]: Spotted trailing whitespace. Removed.

**Comment by cognophile** [comment]
> Overall it looks good! I'm happy to accept if this PR is in a steady state

That's really exciting! I've addressed your whitespace concerns. Unless you've any other feedback, I've nothing else to commit in this PR. I've not bumped any versions as to not presume to know your preference on this. Apologies again for the overgrown PR, but thanks for letting me contribute something and for being welcoming and supportive 🙂

**Pull request merged**